### PR TITLE
ssh key addition for Kubernetes NuvlaEdge

### PR DESCRIPTION
**Pull request by giovannibianco**
initial commit with some debugging statements

### CODE REVIEW

**Comment by giovannibianco** [comment]
OK... We've got the ssh key adding and revocation working now. Ready for demo.

File: code/src/nuvla/connector/kubernetes.py [hunk 1]
```diff
@@ -498,3 +505,178 @@ def extract_vm_ports_mapping(self, vm):
...
+
+    @should_connect
+    def k8s_ssh_key(self, action, pubkey, user_home):
+        '''Doc string'''
+
+        log.debug('We have CA file %s ', self.ca)
+        log.debug('We have certificate file %s ', self.cert)
+        log.debug('We have key file %s ', self.key)
+        log.debug('We have endpoint %s ', self.endpoint)
+        log.debug('We have user home directory %s ', user_home)
```
Comment (schaubl) [inline code/src/nuvla/connector/kubernetes.py:553]: Maybe we can remove the prefix `We have`

File: code/src/nuvla/connector/kubernetes.py [hunk 2]
```diff
@@ -498,3 +505,178 @@ def extract_vm_ports_mapping(self, vm):
...
+                as reboot_manifest_file:
+                reboot_manifest_file.write(formatted_reboot_yaml_manifest)
+            cmd_ssh_key = \
+                self.build_cmd_line(['apply', '-f', tmp_dir_name + '/reboot_job_manifest.yaml'])
+            ssh_key_result = execute_cmd(cmd_ssh_key)
+            log.debug('The result of the ssh key addition : %s',ssh_key_result)
+        return ssh_key_result
+
+    def handle_ssh_key(self, action, pubkey, credential_id, nuvlabox_id):
+        '''Doc string'''
```
Comment (schaubl) [inline code/src/nuvla/connector/kubernetes.py:620]: ?

File: code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py [hunk 3]
```diff
@@ -1,27 +1,25 @@
...
 @action('nuvlabox_add_ssh_key', True)
 class NBAddSSHKey(object):
+    '''Doc string'''
 
     def __init__(self, _, job):
         self.job = job
         self.api = job.api
 
     def add_ssh_key(self):
+        '''Here'''
```
Comment (schaubl) [inline code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py:20]: ?

File: code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py [hunk 4]
```diff
@@ -1,27 +1,25 @@
...
+import re
 
 from ..actions import action
 from ...connector import nuvlabox as NB
-
+from ...connector.kubernetes import K8sSSHKey
 
 @action('nuvlabox_add_ssh_key', True)
 class NBAddSSHKey(object):
+    '''Doc string'''
```
Comment (schaubl) [inline code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py:13]: ?

File: code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py [hunk 5]
```diff
@@ -30,25 +28,42 @@ def add_ssh_key(self):
...
-                r = "Requested SSH key has already been added to the NuvlaBox in the past. Nothing to do..."
         else:
             raise Exception('Cannot find any reference to an existing credential ID')
 
-        self.job.update_job(status_message=json.dumps(r))
-
         return 0
 
+    def _sanitize_ssh_key(self, public_key):
+        '''Doc string'''
```
Comment (schaubl) [inline code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py:63]: ?

File: code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py [hunk 6]
```diff
@@ -30,25 +28,42 @@ def add_ssh_key(self):
...
-
         return 0
 
+    def _sanitize_ssh_key(self, public_key):
+        '''Doc string''' 
+        logging.debug("Checking ssh public key %s ",public_key)
+        return re.match(r'^ssh-\w+\s+\S+(\s+\S+)?$', public_key)
+
     def do_work(self):
+        '''Doc string'''
```
Comment (schaubl) [inline code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py:68]: ?

File: code/src/nuvla/job/actions/nuvlabox_revoke_ssh_key.py [hunk 7]
```diff
@@ -1,24 +1,25 @@
...
 
 from ..actions import action
 from ...connector import nuvlabox as NB
-
+from ...connector.kubernetes import K8sSSHKey
 
 @action('nuvlabox_revoke_ssh_key', True)
 class NBRevokeSSHKey(object):
-
+    '''Here'''
```
Comment (schaubl) [inline code/src/nuvla/job/actions/nuvlabox_revoke_ssh_key.py:13]: ?

File: code/src/nuvla/job/actions/nuvlabox_revoke_ssh_key.py [hunk 8]
```diff
@@ -1,24 +1,25 @@
...
     def __init__(self, _, job):
         self.job = job
         self.api = job.api
 
     def revoke_ssh_key(self):
+        '''Here'''
```
Comment (schaubl) [inline code/src/nuvla/job/actions/nuvlabox_revoke_ssh_key.py:19]: ?

File: container-lite/pom.xml [hunk 9]
```diff
@@ -223,7 +223,6 @@
               <repository>docker.io/${docker-organization}/job-lite</repository>
               <tag>${docker-tag}</tag>
               <executable>${basedir}/container-push.sh</executable>
-              <!-- DOCKER_USERNAME and DOCKER_PASSWORD must be set externally. -->
```
Comment (schaubl) [inline container-lite/pom.xml:226]: Why remove this comment ?

File: container/pom.xml [hunk 10]
```diff
@@ -224,7 +224,6 @@
               <repository>docker.io/${docker-organization}/job</repository>
               <tag>${docker-tag}</tag>
               <executable>${basedir}/container-push.sh</executable>
-              <!-- DOCKER_USERNAME and DOCKER_PASSWORD must be set externally. -->
```
Comment (schaubl) [inline container/pom.xml:227]: Why remove this comment ?

**Review by schaubl — CHANGES_REQUESTED** [review]
In general there are couple of duplicate between `nuvlabox.py` and `kubernetes.py`.
I think this should be refactored later on in a dedicated task.
Like in NuvlaEdge code, the NuvlaBox class should have two coe class (docker and kubernetes) so that the common part will be on the NuvlaBox class.

Has this branch been tested (ssh add&delete) with a Docker based NuvlaEdge ?

File: code/src/nuvla/connector/kubernetes.py [hunk 11]
```diff
@@ -20,12 +21,17 @@ def instantiate_from_cimi(api_infrastructure_service, api_credential):
...
+    kubernetes_port = os.getenv('KUBERNETES_SERVICE_PORT')
+    if kubernetes_host and kubernetes_port:
+        return f'https://{kubernetes_host}:{kubernetes_port}'
+    return ''
 class Kubernetes(Connector):
 
     def __init__(self, **kwargs):
         super(Kubernetes, self).__init__(**kwargs)
-
+        '''doc string'''
```
[the remark this answers is not in the document]
Comment (giovannibianco) [reply code/src/nuvla/connector/kubernetes.py:39]: OK, so now I see that we do not doc string these function definitions within these classes.
I will conform to the overall practice and remove this doc string.

File: code/src/nuvla/connector/kubernetes.py [hunk 12]
```diff
@@ -58,6 +64,7 @@ def clear_connection(self, connect_result):
             self.key_file = None
 
     def build_cmd_line(self, list_cmd):
+        '''Doc string'''
```
[the remark this answers is not in the document]
Comment (giovannibianco) [reply code/src/nuvla/connector/kubernetes.py:72]: For these smaller function definitions I see that a doc string is appropriate.
Added text.

File: code/src/nuvla/connector/kubernetes.py [hunk 13]
```diff
@@ -498,3 +505,178 @@ def extract_vm_ports_mapping(self, vm):
...
+        """ Updates the NuvlaEdge resource with the provided payload
+        :param payload: content to be updated in the NuvlaEdge resource
+        """
+        # pass
+        if payload:
+            self.api.operation(self.nuvlabox_resource, "commission", data=payload)
+
+
+class K8sSSHKey(Kubernetes):
+    '''Here'''
```
[the remark this answers is not in the document]
Comment (giovannibianco) [reply code/src/nuvla/connector/kubernetes.py:513]: Added a doc string to describe this class function.

File: code/pom.xml [hunk 14]
```diff
@@ -96,6 +96,7 @@
               <goal>exec</goal>
             </goals>
             <configuration>
+              <!-- executable>/Users/jwhite/Library/Python/3.9/bin/pytest</executable -->
```
[the remark this answers is not in the document]
Comment (giovannibianco) [reply code/pom.xml:99]: Agree. Removed.

File: code/src/nuvla/connector/kubernetes.py (hunk 1 above)
Comment (giovannibianco) [reply code/src/nuvla/connector/kubernetes.py:547]: OK... mainly for my readability in debugging. Removed.

File: code/src/nuvla/connector/kubernetes.py (hunk 2 above)
Comment (giovannibianco) [reply code/src/nuvla/connector/kubernetes.py:614]: Added an appropriate doc string.

File: code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py (hunk 3 above)
Comment (giovannibianco) [reply code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py:20]: Added an appropriate doc string.

File: code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py (hunk 4 above)
Comment (giovannibianco) [reply code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py:13]: Added an appropriate doc string.

File: code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py (hunk 5 above)
Comment (giovannibianco) [reply code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py:63]: Added an appropriate doc string.

File: code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py (hunk 6 above)
Comment (giovannibianco) [reply code/src/nuvla/job/actions/nuvlabox_add_ssh_key.py:68]: Removed the doc string to conform to how other do_work are handled

File: code/src/nuvla/job/actions/nuvlabox_revoke_ssh_key.py (hunk 7 above)
Comment (giovannibianco) [reply code/src/nuvla/job/actions/nuvlabox_revoke_ssh_key.py:13]: Added an appropriate doc string.

File: code/src/nuvla/job/actions/nuvlabox_revoke_ssh_key.py (hunk 8 above)
Comment (giovannibianco) [reply code/src/nuvla/job/actions/nuvlabox_revoke_ssh_key.py:19]: Removed the doc string to conform to how other do_work are handled

File: container-lite/pom.xml (hunk 9 above)
Comment (giovannibianco) [reply container-lite/pom.xml:226]: Probably removed by mistake when adjusting this file for local build. 
Restored.

File: .gitignore
[the remark this answers is not in the document]
Comment (giovannibianco) [reply .gitignore:1]: Hmm... this was probably a temporary change that got forgotten.
Restored.

File: code/src/nuvla/connector/kubernetes.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (giovannibianco) [reply code/src/nuvla/connector/kubernetes.py:39]: I see that there are no doc strings for same definition in other classes.
Added the doc string to keep SonarCloud quiet.
Removed the doc string.

File: .gitignore
[the remark this answers is not in the document]
Comment (giovannibianco) [reply .gitignore:1]: This was a temporary change that got forgotten.
Restored.

**Comment by giovannibianco** [comment]
I have addressed all the comments in the above review...

**Review by schaubl — APPROVED** [review]

**Pull request merged**